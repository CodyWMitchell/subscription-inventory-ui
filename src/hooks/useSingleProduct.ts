import { useQuery, QueryObserverResult } from 'react-query';
import { Product } from './useProducts';
import Cookies from 'js-cookie';

interface SingleProductApiData {
  body: Product;
}

const fetchProductData = async (sku: string): Promise<Product> => {
  const jwtToken = Cookies.get('cs_jwt');

  const response = await fetch(`/api/rhsm/v2/products/${sku}`, {
    headers: { Authorization: `Bearer ${jwtToken}` }
  });

  const productResponseData: SingleProductApiData = await response.json();

  return productResponseData.body;
};

const getSingleProduct = async (sku: string): Promise<Product> => {
  const productData = await fetchProductData(sku);

  return productData;
};

const useSingleProduct = (sku: string): QueryObserverResult<Product, unknown> => {
  return useQuery(`singleProduct.${sku}`, () => getSingleProduct(sku));
};

export { useSingleProduct as default };
