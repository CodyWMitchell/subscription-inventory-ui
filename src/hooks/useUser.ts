import { useQuery, UseQueryResult } from 'react-query';
import Cookies from 'js-cookie';
import { getConfig, authenticateUser, getUserRbacPermissions } from '../utilities/platformServices';

interface User {
  accountNumber: string;
  canReadProducts: boolean;
  isOrgAdmin: boolean;
}

const getUser = (): Promise<User> => {
  return Promise.all([authenticateUser(), getUserRbacPermissions()]).then(
    ([userStatus, rawRbacPermissions]) => {
      const rbacPermissions = rawRbacPermissions.map((rawPermission) => rawPermission.permission);
      const user: User = {
        accountNumber: userStatus.identity.account_number,
        canReadProducts:
          rbacPermissions.includes('subscriptions:products:read') ||
          rbacPermissions.includes('subscriptions:*:*'),
        isOrgAdmin: userStatus.identity.user.is_org_admin === true
      };
      return user;
    }
  );
};

const useUser = (): UseQueryResult<User, unknown> => {
  return useQuery('user', () => getUser());
};

export { getUser, useUser as default, User };
