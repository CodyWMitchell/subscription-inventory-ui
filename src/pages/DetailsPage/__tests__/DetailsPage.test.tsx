import React from 'react';
import { render, waitFor } from '@testing-library/react';
import DetailsPage from '../DetailsPage';
import Authentication from '../../../components/Authentication';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { init } from '../../../store';
import { QueryClient, QueryClientProvider } from 'react-query';
import useUser from '../../../hooks/useUser';
import useSingleProduct from '../../../hooks/useSingleProduct';

jest.mock('../../../hooks/useUser');
jest.mock('../../../hooks/useSingleProduct');
jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as Record<string, unknown>),
  useLocation: () => ({
    pathname: '/'
  }),
  useParams: () => ({
    SKU: 'TESTSKU'
  })
}));

const queryClient = new QueryClient();

const Page = () => (
  <QueryClientProvider client={queryClient}>
    <Authentication>
      <Provider store={init().getStore()}>
        <Router>
          <DetailsPage />
        </Router>
      </Provider>
    </Authentication>
  </QueryClientProvider>
);

const mockAuthenticateUser = (
  isLoading: boolean,
  orgAdminStatus: boolean,
  canReadProducts: boolean
) => {
  (useUser as jest.Mock).mockReturnValue({
    isLoading: isLoading,
    isFetching: false,
    isSuccess: true,
    isError: false,
    data: {
      isOrgAdmin: orgAdminStatus,
      isSCACapable: true,
      canReadProducts
    }
  });

  queryClient.setQueryData('user', {
    isSCACapable: true,
    isOrgAdmin: orgAdminStatus,
    canReadProducts
  });
};

const mockSingleProduct = (hasUom: boolean) => {
  (useSingleProduct as jest.Mock).mockReturnValue({
    isLoading: false,
    isFetching: false,
    isSuccess: true,
    isError: false,
    data: {
      name: 'TEST Name',
      productLine: 'TEST Line',
      quantity: 3,
      sku: 'TESTSKU',
      serviceLevel: '',
      serviceType: '',
      unitOfMeasure: hasUom ? { name: 'test', quantity: 2 } : null
    }
  });

  queryClient.setQueryData('singleProduct.TESTSKU', {
    name: 'TEST Name',
    productLine: 'TEST Line',
    quantity: 3,
    sku: 'TESTSKU',
    serviceLevel: 'TEST serviceLevel',
    serviceType: 'TEST serviceType',
    unitOfMeasure: hasUom ? { name: 'test', quantity: 2 } : null
  });
};

describe('Details Page', () => {
  it('loader shows correctly', async () => {
    window.insights = {};
    const isLoading = true;
    const isOrgAdmin = true;
    mockAuthenticateUser(isLoading, isOrgAdmin, true);

    const { container } = render(<Page />);

    await waitFor(() => expect(useUser).toHaveBeenCalledTimes(1));
    expect(container).toMatchSnapshot();
  });

  it('renders data', async () => {
    const isLoading = false;
    const isOrgAdmin = true;
    const canReadProducts = true;
    mockAuthenticateUser(isLoading, isOrgAdmin, canReadProducts);
    mockSingleProduct(true);

    const { container } = render(<Page />);

    await waitFor(() => expect(useSingleProduct).toHaveBeenCalledTimes(1));
    expect(container).toMatchSnapshot();
  });

  it("redirects when can't read products", async () => {
    const isLoading = false;
    const isOrgAdmin = true;
    const canReadProducts = false;
    mockAuthenticateUser(isLoading, isOrgAdmin, canReadProducts);
    mockSingleProduct(false);

    const { container } = render(<Page />);

    await waitFor(() => expect(useSingleProduct).toHaveBeenCalledTimes(2));
    expect(container).toMatchSnapshot();
  });

  it('renders not available for missing data', async () => {
    const isLoading = false;
    const isOrgAdmin = true;
    const canReadProducts = true;
    mockAuthenticateUser(isLoading, isOrgAdmin, canReadProducts);
    mockSingleProduct(false);

    const { container } = render(<Page />);

    await waitFor(() => expect(useSingleProduct).toHaveBeenCalledTimes(1));
    expect(container).toMatchSnapshot();
  });
});
