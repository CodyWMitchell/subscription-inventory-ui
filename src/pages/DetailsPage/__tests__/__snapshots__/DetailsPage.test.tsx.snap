// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Page loader shows correctly 1`] = `
<div>
  <div
    class="pf-l-bullseye"
  >
    <span
      aria-label="Contents"
      aria-valuetext="Loading..."
      class="pf-c-spinner pf-m-xl"
      role="progressbar"
    >
      <span
        class="pf-c-spinner__clipper"
      />
      <span
        class="pf-c-spinner__lead-ball"
      />
      <span
        class="pf-c-spinner__tail-ball"
      />
    </span>
  </div>
</div>
`;

exports[`Details Page redirects when can't read products 1`] = `<div />`;

exports[`Details Page renders data 1`] = `
<div>
  <section
    class="pf-l-page-header pf-c-page-header pf-l-page__main-section pf-c-page__main-section pf-m-light"
    widget-type="InsightsPageHeader"
  >
    <nav
      aria-label="Breadcrumb"
      class="pf-c-breadcrumb"
      data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
      data-ouia-component-type="PF4/Breadcrumb"
      data-ouia-safe="true"
    >
      <ol
        class="pf-c-breadcrumb__list"
      >
        <li
          class="pf-c-breadcrumb__item"
        >
          <a
            href="/"
          >
            Subscription Inventory
          </a>
        </li>
        <li
          class="pf-c-breadcrumb__item"
        >
          <span
            class="pf-c-breadcrumb__item-divider"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 256 512"
              width="1em"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
              />
            </svg>
          </span>
          TEST Name
           
        </li>
      </ol>
    </nav>
    <h1
      class="pf-c-title pf-m-2xl"
      data-ouia-component-id="OUIA-Generated-Title-1"
      data-ouia-component-type="PF4/Title"
      data-ouia-safe="true"
      widget-type="InsightsPageHeaderTitle"
    >
      TEST Name
    </h1>
    <ul
      class="pf-c-list pf-m-plain"
    >
      <li
        class="pf-u-mt-md"
      >
        <b>
          SKU: 
        </b>
        TESTSKU
      </li>
      <li
        class="pf-u-mt-0"
      >
        <b>
          Quantity: 
        </b>
        3
      </li>
      <li
        class="pf-u-mt-0"
      >
        <b>
          Support level: 
        </b>
        TEST serviceLevel
      </li>
      <li
        class="pf-u-mt-0"
      >
        <b>
          Support type: 
        </b>
        TEST serviceType
      </li>
      <li
        class="pf-u-mt-0"
      >
        <b>
          Capacity: 
        </b>
        test
         
        <span
          class="pf-c-badge pf-m-read"
        >
          2
        </span>
      </li>
    </ul>
  </section>
</div>
`;

exports[`Details Page renders not available for missing data 1`] = `
<div>
  <section
    class="pf-l-page-header pf-c-page-header pf-l-page__main-section pf-c-page__main-section pf-m-light"
    widget-type="InsightsPageHeader"
  >
    <nav
      aria-label="Breadcrumb"
      class="pf-c-breadcrumb"
      data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
      data-ouia-component-type="PF4/Breadcrumb"
      data-ouia-safe="true"
    >
      <ol
        class="pf-c-breadcrumb__list"
      >
        <li
          class="pf-c-breadcrumb__item"
        >
          <a
            href="/"
          >
            Subscription Inventory
          </a>
        </li>
        <li
          class="pf-c-breadcrumb__item"
        >
          <span
            class="pf-c-breadcrumb__item-divider"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 256 512"
              width="1em"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
              />
            </svg>
          </span>
          
           
        </li>
      </ol>
    </nav>
    <h1
      class="pf-c-title pf-m-2xl"
      data-ouia-component-id="OUIA-Generated-Title-1"
      data-ouia-component-type="PF4/Title"
      data-ouia-safe="true"
      widget-type="InsightsPageHeaderTitle"
    />
    <ul
      class="pf-c-list pf-m-plain"
    >
      <li
        class="pf-u-mt-md"
      >
        <b>
          SKU: 
        </b>
        Not Available
      </li>
      <li
        class="pf-u-mt-0"
      >
        <b>
          Quantity: 
        </b>
        0
      </li>
      <li
        class="pf-u-mt-0"
      >
        <b>
          Support level: 
        </b>
        Not Available
      </li>
      <li
        class="pf-u-mt-0"
      >
        <b>
          Support type: 
        </b>
        Not Available
      </li>
      <li
        class="pf-u-mt-0"
      >
        <b>
          Capacity: 
        </b>
        Not Available
      </li>
    </ul>
  </section>
</div>
`;
